include/deal.II/fe/mapping_fe.h

```diff
@@ -0,0 +1,741 @@
+
+  /**
+   * Implementation of the interface in Mapping.
+   */
+  virtual void
+  fill_fe_subface_values (const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                          const unsigned int face_no,
+                          const unsigned int sub_no,
+                          const Quadrature<dim-1>& quadrature,
+                          typename Mapping<dim,spacedim>::InternalDataBase &mapping_data,
+                          typename std::vector<Point<spacedim> >        &quadrature_points,
+                          std::vector<double>             &JxW_values,
+                          std::vector<Tensor<1,spacedim> > &exterior_forms,
+                          std::vector<Point<spacedim> >    &normal_vectors,
+                          std::vector<DerivativeForm<1,dim,spacedim> > &jacobians,
+                          std::vector<DerivativeForm<1,spacedim,dim> > &inverse_jacobians) const ;
+
+
+  /**
+     This function and the next allow to generate the transform require by
+     the virtual transform() in mapping, but unfortunately in C++ one cannot
+     declare a virtual template function.
+  */
+  template < int rank >
+  void
+  transform_fields(const VectorSlice<const std::vector<Tensor<rank,dim>      > > input,
+                   VectorSlice<      std::vector<Tensor<rank,spacedim> > > output,
+                   const typename Mapping<dim,spacedim>::InternalDataBase &internal,
+                   const MappingType type) const;
+
+
+  /**
+       see doc in transform_fields
+   */
+  template < int rank >
+  void
+  transform_differential_forms(
+    const VectorSlice<const std::vector<DerivativeForm<rank, dim,spacedim> > >    input,
+    VectorSlice<std::vector<DerivativeForm<rank, spacedim,spacedim> > > output,
+    const typename Mapping<dim,spacedim>::InternalDataBase &mapping_data,
+    const MappingType mapping_type) const;
+
+
+protected:
+  /**
+  * Reference to the vector of shifts.
+  */
+
+  SmartPointer<const VECTOR, MappingFE<dim,spacedim,DH,VECTOR> >euler_vector;
+  /**
+   * A FiniteElement object which is only needed in 3D, since it knows how to reorder
+   * shape functions/DoFs on non-standard faces. This is used to reorder
+   * support points in the same way. We could make this a pointer to prevent
+   * construction in 1D and 2D, but since memory and time requirements are not
+   * particularly high this seems unnecessary at the moment.
+   */
+  SmartPointer<const FiniteElement<dim,spacedim>, MappingFE<dim,spacedim,DH,VECTOR> > fe;
+
+
+  /**
+   * Pointer to the DoFHandler to which the mapping vector is associated.
+   */
+  SmartPointer<const DH,MappingFE<dim,spacedim,DH,VECTOR> >euler_dof_handler;
+
+
+
+private:
+//
+  /**
+   * Update internal degrees of
+   * freedom. */
+  void update_internal_dofs(const typename Triangulation<dim,spacedim>::cell_iterator &cell) const;
+
+
+  mutable std::vector<double> local_dofs;
+
+  mutable std::vector<unsigned int> dof_indices;
```

Comment:
By the way, this doesn't currently compile at all when in 64-bit mode. The correct data type should be 
```
  mutable std::vector<types::global_dof_index> dof_indices;
```